packages/drivers/odsp-driver/src/odspSummaryUploadManager.ts

```diff
@@ -0,0 +1,412 @@
+     *  c.) Building new trees/blobs dedup caches so that they can be used to dedup blobs in next summary.
+     * @param parentHandle - Handle of the last uploaded summary or detach new summary.
+     * @param tree - Summary Tree which will be converted to snapshot tree to be uploaded.
+     * @param blobTreeDedupCachesLatest - Blobs/Trees caches which will be build to be used in next summary upload
+     *  in order to dedup the blobs.
+     * @param rootNodeName - Root node name of the summary tree.
+     * @param path - Current path of node which is getting evaluated.
+     * @param expanded - True if we are currently expanding a handle by a tree stored in the cache.
+     */
+    private async convertSummaryToSnapshotTree(
+        parentHandle: string | undefined,
+        tree: api.ISummaryTree,
+        blobTreeDedupCachesLatest: IDedupCaches,
+        rootNodeName: string,
+        path: string = "",
+        expanded: boolean = false,
+    ) {
+        const snapshotTree: ISnapshotTree = {
+            entries: [],
+        }!;
+
+        let reusedBlobs = 0;
+        let blobs = 0;
+
+        const keys = Object.keys(tree.tree);
+        for (const key of keys) {
+            const summaryObject = tree.tree[key];
+
+            let id: string | undefined;
+            let value: SnapshotTreeValue | undefined;
+            const currentPath = path === "" ? key : `${path}/${key}`;
+            switch (summaryObject.type) {
+                case api.SummaryType.Tree: {
+                    blobTreeDedupCachesLatest.treesPathToTree.set(currentPath, summaryObject);
+                    const result = await this.convertSummaryToSnapshotTree(
+                        parentHandle,
+                        summaryObject,
+                        blobTreeDedupCachesLatest,
+                        rootNodeName,
+                        currentPath,
+                        expanded);
+                    value = result.snapshotTree;
+                    reusedBlobs += result.reusedBlobs;
+                    blobs += result.blobs;
+                    break;
+                }
+                case api.SummaryType.Blob: {
+                    let hash: string | undefined;
+                    let cachedPath: string | undefined;
+                    // If we are expanding the handle, then the currentPath should exist in the cache as we will get the blob
+                    // hash from the cache.
+                    if (expanded) {
+                        hash = this.blobTreeDedupCaches.pathToBlobSha.get(currentPath);
+                        assert(hash !== undefined, "hash should be set here");
+                        cachedPath = this.blobTreeDedupCaches.blobShaToPath.get(hash);
+                        assert(cachedPath !== undefined, "path should be defined as path->sha mapping exists");
+                    } else {
+                        value = typeof summaryObject.content === "string"
```

Comment:
If I got it right, a blob content in the tree will stay here and will make it (as part the tree) into blobTreeDedupCachesLatest.treesPathToTree. In other words, we keep blob content across summaries.
I think that needs to be addressed - we should not keep any memory around other than shallow data (tree structure, paths, shas)

Reply:
Is there any way we can write UT to validate that (how much memory we use / blobs being not kept around)?

Reply:
Now we just store empty strings as content for blobs. 
You mean how much we save per blobs? Even if we write that it will just depend on the number of blobs and content lengths that we will put in that test.